File: Data Structures & Algorithms/rotting-fruit/submission-2.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int ROWS = ssize(grid), COLS = ssize(grid[0]);
        queue<pair<int, int>> q;
        const vector<pair<int, int>> neighbor_differences = {
            {-1, 0}, {0, 1}, {1, 0}, {0, -1}
        };

        int fresh_count = 0;
        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                if (grid[r][c] == 2) q.push({r, c});
                if (grid[r][c] == 1) fresh_count++;
            }
        }

        int minutes = 0;
        while (!q.empty() && fresh_count > 0) {
            int q_size = ssize(q);
            for (int i = 0; i < q_size; i++) {
                auto [r, c] = q.front();
                q.pop();
                for (const pair<int, int>& neighbor_difference : neighbor_differences) {
                    int adjRow = r + neighbor_difference.first;
                    int adjCol = c + neighbor_difference.second;
                    if (min(adjRow, adjCol) < 0 || 
                        adjRow == ROWS || 
                        adjCol == COLS || 
                        grid[adjRow][adjCol] == 2 || 
                        grid[adjRow][adjCol] == 0) continue;
                    
                    q.push({adjRow, adjCol});
                    grid[adjRow][adjCol] = 2;
                    fresh_count--;
                }
            }
            minutes++;
        }
        return fresh_count > 0 ? -1 : minutes;
    }
};

File: Data Structures & Algorithms/islands-and-treasure/submission-2.cs
public class Solution {
    public void islandsAndTreasure(int[][] grid) {
        int ROWS = grid.Length, COLS = grid[0].Length;
        var queue = new Queue<(int r, int c)>();
        (int r, int c)[] neighbors = {
            (-1, 0), (0, 1), (1, 0), (0, -1)
        };

        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                if (grid[r][c] == 0) queue.Enqueue((r, c));
            }
        }

        int length = 1;
        while (queue.Count > 0) {
            int queueSize = queue.Count;
            for (int i = 0; i < queueSize; i++) {
                (int r, int c) = queue.Dequeue();

                foreach ((int dr, int dc) in neighbors) {
                    int adjRow = r + dr, adjCol = c + dc;
                    if (Math.Min(adjRow, adjCol) < 0 || 
                        adjRow == ROWS || 
                        adjCol == COLS || 
                        grid[adjRow][adjCol] != int.MaxValue) continue;
                    
                    queue.Enqueue((adjRow, adjCol));
                    grid[adjRow][adjCol] = length;
                }
            }
            length++;
        }
    }
}
